File: java-code/src/main/java/refactoring/split_phase/SplitPhase_After.java
package refactoring.split_phase;

public class SplitPhase_After {
  
  public double priceOrder(Product product, int quantity, ShippingMethod shippingMethod) {

    //Step4: Move the pricing logic also to a separate method and let it return intermediate object
    PriceData priceData = calculatePricingData(product, quantity);

    return applyShipping(priceData, shippingMethod);
  }

  private PriceData calculatePricingData(Product product, int quantity) {
    double basePrice = product.getBasePrice() * quantity;
    double discount =
        Math.max(quantity - product.getDiscountThreshold(), 0)
            * product.getBasePrice()
            * product.getDiscountRate();

    // Step 2: Create an intermediate class
    // Step 3: Move the properties into this intermediate class
    return new PriceData(basePrice, quantity, discount);
  }

  // Step 1: Extract shipping Logic
  private double applyShipping(
      PriceData priceData, ShippingMethod shippingMethod) {
    double shippingPerCase =
        (priceData.getBasePrice() > shippingMethod.getDiscountThreshold())
            ? shippingMethod.getDiscountedFee()
            : shippingMethod.getFeePerCase();
    double shippingCost = priceData.getQuantity() * shippingPerCase;
    return priceData.getBasePrice() - priceData.getDiscount() + shippingCost;
  }
}

File: go-code/basics/split_phase/split_phase_after.go
package split_phase

import "math"

// PriceData Step 2: Create intermediate struct
type PriceData struct {

	//Step3: Move pricing related data into intermediate struct
	Qantity   int
	BasePrice float64
	Discount  float64
}

func priceOrderAfter(product Product, quantity int, shippingMethod ShippingMethod) float64 {

	//Step 4: Move price calculation phase into a separate method
	priceData := calculatePricingData(product, quantity)

	return applyShipping(priceData, shippingMethod)
}

func calculatePricingData(product Product, quantity int) PriceData {
	basePrice := product.BasePrice * float64(quantity)
	discount := math.Max(float64(quantity-product.DiscountThreshold), 0) * product.BasePrice * product.DiscountRate

	priceData := PriceData{
		BasePrice: basePrice,
		Qantity:   quantity,
		Discount:  discount,
	}
	return priceData
}

// Step 1: split the shipping part from the main func
func applyShipping(priceData PriceData, shippingMethod ShippingMethod) float64 {
	shippingPerCase := shippingMethod.FeePerCase
	if priceData.BasePrice > float64(shippingMethod.DiscountThreshold) {
		shippingPerCase = shippingMethod.DiscountedFee
	}
	shippingCost := float64(priceData.Qantity) * shippingPerCase

	return priceData.BasePrice - priceData.Discount + shippingCost
}
